src/v/kafka/security/scram_algorithm.h

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+#include "bytes/bytes.h"
+#include "hashing/secure.h"
+#include "kafka/logger.h"
+#include "random/generators.h"
+
+#include <absl/container/node_hash_map.h>
+
+#include <regex>
+
+/**
+ * scram algorthm - https://tools.ietf.org/html/rfc5802
+ *
+ * The implementation follows closely the rfc, with some scenarios and
+ * configuration not handled by kafka omitted from the implementation. The kafka
+ * implementation is closely based on the rfc as well, so they are fairly easy
+ * to compare.
+ *
+ * TODO
+ * ====
+ *
+ * 1. unnecessary copying from std::array<char> for hash digest types and
+ * built-in bytes type. the conversion is mostly mechanical, but std::array
+ * needs to be propogated to all other types which need to be templated on the
+ * same scram_algorithm template parameters.
+ */
+namespace kafka {
+
+class scram_exception final : public std::exception {
+public:
+    explicit scram_exception(ss::sstring msg) noexcept
+      : _msg(std::move(msg)) {}
+
+    const char* what() const noexcept final { return _msg.c_str(); }
+
+private:
+    ss::sstring _msg;
+};
+
+class scram_credential {
+public:
+    scram_credential() noexcept = default;
+
+    scram_credential(
+      bytes salt, bytes server_key, bytes stored_key, int iterations) noexcept
+      : _salt(std::move(salt))
+      , _server_key(std::move(server_key))
+      , _stored_key(std::move(stored_key))
+      , _iterations(iterations) {}
+
+    const bytes& salt() const { return _salt; }
+    const auto& server_key() const { return _server_key; }
```

Comment:
what's the point of this auto ?  - we don't have this kinda code anywhere else.... why not bytes& also, mixed style is confusing

Reply:
its a mistake

Reply:
fixed